onnxruntime/core/providers/dml/DmlExecutionProvider/src/TensorDesc.cpp

```diff
@@ -315,3 +315,52 @@
     }
     m_bufferTensorDesc.DimensionCount = newDimensionCount;
 }
+
+// Uses dimensionMapping to reorder m_sizes and m_strides to match specific Tensor layout
+void TensorDesc::PermuteDimensions(const std::vector<uint32_t> dimensionMapping, const TensorAxis alignment)
+{
+    EnsureMinimumDimensionCount(dimensionMapping.size(), alignment);
+    InitializeStrides(static_cast<int>(dimensionMapping.size()), alignment);
+    PermuteArray(dimensionMapping, alignment);
+}
+
+// Shuffle m_sizes and m_strides acording to the indexes pointed by dimensionMapping
+void TensorDesc::PermuteArray(const std::vector<uint32_t> dimensionMapping, const TensorAxis alignment)
```

Comment:
>PermuteArray [](http://example.com/codeflow?start=17&length=12)

We probably don't want this function to be publicly callable outside TensorDesc, as it could introduce some weirdness and unsafety (because `PermuteDimensions` and `PermuteArray` are identical in signature, but one might lead to badness later because the dimension count and strides were not set properly 😉).

I'd just make it a little helper in an anonymous namespace that takes a single array (in/out parameter) and permutes it.

Reply:
🤔 With such a helper, then this becomes much lighter, and most functions don't even need to know about the alignment, since SetDimensionCount already hands it. e.g...


```
// Uses dimensionMapping to reorder m_sizes and m_strides to match specific Tensor layout
void TensorDesc::PermuteDimensions(gsl::span<const uint32_t> dimensionMapping, TensorAxis alignment)
{
   uint32_t newDimensionCount = ...
    EnsureStridesInitialized();
    EnsureMinimumDimensionCount(newDimensionCount , alignment);

    PermuteArray(/*inout*/ gsl::span(std::data(m_sizes), newDimensionCount ), dimensionMapping);
    PermuteArray(/*inout*/ gsl::span(std::data(m_strides), newDimensionCount ), dimensionMapping);

    m_bufferTensorDesc.Sizes = m_sizes;
    m_bufferTensorDesc.Strides = m_strides;
}
```

Reply:
`assert(values.size() == dimensionMapping.size());`
Would this be true for Right Aligned Tenors?

Reply:
I see two options `PermuteArray` as a helper function which also needs `TensorAxis alignment`, and making `PermuteArray`(and also optionally rename as `PermuteArrayHelper`) private so that its not available for public use to avoid some confution

Reply:
> Would this be true for Right Aligned Tenors? 

Alignment just sets how the original dimensions are aligned. Both left and right alignment and have the same number of dimensions

Reply:
> PermuteArray as a helper function which also needs TensorAxis alignment

Since a new size has already been set and adjusted by the time this function is called, the TensorAxis is no longer relevant.

> making PermuteArray private so that its not available for public use to avoid some confusion

Yes, at least not public either way. Though, unless a function *requires* class state, it's often better (because it's easier to reason about all the inputs and outputs at a glance, and it's more flexible/generic) to just make it a non-class helper function.

> Rename as PermuteArrayHelper or PermuteSizeAndStrides 

That would be clearer. Though, you're still left with the weird split of responsibilities between the two functions `PermuteDimensions` and `PermuteSizeAndStrides`, where both take the exact same parameters, and `PermuteSizeAndStrides` does 90% of the work that `PermuteDimensions` needs to do, and `PermuteSizeAndStrides` probably won't ever be called separately by itself. So as-is, they could just be combined.

Reply:
moved PermuteArray content into PermuteDimensions in follow up revision